### DISCUSSION TITLE
Compatibility

### DISCUSSION THREAD

**Discussion by juliosteinlager** [ksheumaker/homeassistant-apsystems_ecur#130] [Ideas]
I have an ECU with an ID starting with 203000 
I have not been able to get it communicating with this integration, and by the  instructions that is to be expected. (unless I miss understood a bunch). Is there anything I can do to assist in adding support for this unit?

**Comment by HAEdwin** [comment]
Tell me more about your ECU (model, type, photo). Until now we have the following ECU's:
216xxxxxxxxx = ECU-R
2162xxxxxxxx = ECU-R SunSpec
2163xxxxxxxx = ECU-B
215xxxxxxxxx = ECU-C

**Reply by juliosteinlager** [reply]
Model is ECU-3 


![IMG_20221104_111900.jpg](https://user-images.githubusercontent.com/64613171/200048517-d47d045e-cddc-40fa-89ec-bcb62bef17ef.jpg)

![IMG_20221104_111943.jpg](https://user-images.githubusercontent.com/64613171/200048656-200bc007-070f-4b5f-8a13-ef8e30ae9ceb.jpg)

![IMG_20221104_111950.jpg](https://user-images.githubusercontent.com/64613171/200048790-fa57c2d9-1409-408d-9b83-5aa4973ffe2f.jpg)

**Reply by HAEdwin** [reply]
Unfortunately, the ECU-3 v4 is not supported by this integration. The operation is based on an available port that can be queried. This port (8899) is probably not available on the ECU-3 v4 as the integration is not working.
In the example below you see the basis of how the functionality works:
![Use_Packet_Sender](https://user-images.githubusercontent.com/82239374/200058799-248e7a3c-fe53-473d-a1d4-211335c58eeb.jpg)

The command APS1100160001END is being send to the IP Address of the ECU on port 8899. The ECU then responds back with data. You could try other ports using packet sender or an initially port scan to see which ports are open on the ECU-3 v4. Chances are slim that it will work though, sorry...

**Comment by juliosteinlager** [comment]
`This error originated from a custom integration.

Logger: custom_components.apsystems_ecur.config_flow
Source: custom_components/apsystems_ecur/APSystemsSocket.py:104
Integration: APSystems PV solar ECU (documentation, issues)
First occurred: 12:25:59 PM (1 occurrences)
Last logged: 12:25:59 PM

APSystemsInvalidData exception: [Errno 111] Connection refused
Traceback (most recent call last):
  File "/config/custom_components/apsystems_ecur/APSystemsSocket.py", line 101, in open_socket
    self.sock.connect((self.ipaddr, self.port))
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/apsystems_ecur/config_flow.py", line 46, in async_step_user
    test_query = await self.hass.async_add_executor_job(ap_ecu.query_ecu)
  File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/apsystems_ecur/APSystemsSocket.py", line 107, in query_ecu
    self.open_socket()
  File "/config/custom_components/apsystems_ecur/APSystemsSocket.py", line 104, in open_socket
    raise APSystemsInvalidData(err)
custom_components.apsystems_ecur.APSystemsSocket.APSystemsInvalidData: [Errno 111] Connection refused
`

**Reply by HAEdwin** [reply]
Yes, this confirms that port 8899 is not open for communication.

**Reply by juliosteinlager** [reply]
I will fire up a port scanner when I have time and see what is happing there. Thank you.

**Reply by juliosteinlager** [reply]
I ran the port scan and I am showing TPC 4540 and 4560 open as well as the expected port 80.

**Reply by HAEdwin** [reply]
Any responses on those ports when you send the APS command to them?

**Comment by HAEdwin** [comment]
Unfortunately, this integration is not compatible with the ECU-3. It's the way the ECU-3 is installed that is the culprit here.
